### PR TITLE
Extended IRPropMin with boundary support and methods to set min and max step size

### PR DESCRIPTION
**Pull request by MichaelClerx**
⬆️  that

The boundary support is non-standard and can cause you to get stuck if you have a problematic problem (i.e. gradient next to the boundaries pointing outside). But very useful if you need to avoid evaluating outside some range that causes errors...

### CODE REVIEW

**Comment by MichaelClerx** [comment]
@DavAug maybe don't review just yet: I think i'll need to change the strategy for step-size shrinking with general boundaries

**Comment by MichaelClerx** [comment]
OK ready now @DavAug 

- Added max step size
- Added get/set for max & min step size, although default min step size still set with sigma0
- Added boundary checks: for rectangular boundaries this is straightforward. For non-rectangular it's a bit weird: I first reduce all step sizes until the boundaries are OK. Then I try setting them all back to their original values again (one by one). The idea here is that if I need to reduce step sizes in dimensions 3, 7, and 8, then that will take me forever to find out if I start by trying to reduce just 1 step size, then all combinations of 2 step sizes, then all of 3. So this looks stupid but should work

**Comment by MichaelClerx** [comment]
(The strategy is biased against parameters/dimensions with a lower indice: these will get reduced first (even if reducing another parameter would also have worked). I thought a bit about randomisation but that makes this a stochastic method and I don't want to go there. The bias shouldn't matter, I hope: as long as we get to the right solution in the end)

File: pints/_optimisers/_irpropmin.py [hunk 1]
```diff
@@ -127,12 +166,35 @@ def needs_sensitivities(self):
...
+        """
+        self.set_min_step_size(x[0])
+        self.set_max_step_size(x[1])
+
+    def set_max_step_size(self, step_size):
+        """
+        Sets the maximum step size (use ``None`` to let step sizes grow
+        indefinitely).
+        """
+        self._step_max = None if step_size is None else float(step_size)
```
Comment (DavAug) [inline pints/_optimisers/_irpropmin.py:189]: Can we throw an error when step_max < step_min? Does the method break when step_max = step_min?

File: pints/_optimisers/_irpropmin.py [hunk 2]
```diff
@@ -127,12 +166,35 @@ def needs_sensitivities(self):
...
+
+    def set_min_step_size(self, step_size):
+        """
+        Sets the minimum step size (use ``None`` to let step sizes shrink
+        indefinitely).
+        """
+        self._step_min = None if step_size is None else float(step_size)
```
Comment (DavAug) [inline pints/_optimisers/_irpropmin.py:196]: Same here. Checking step_max > step_min is probably a good idea, and that step_min > 0

File: pints/_optimisers/_irpropmin.py [hunk 3]
```diff
@@ -178,7 +238,50 @@ def tell(self, reply):
         self._current_df = dfx
 
         # Take step in direction indicated by current gradient
-        self._proposed = self._current - self._step_size * np.sign(dfx)
+        p = self._current - self._step_size * np.sign(dfx)
+
+        # Allow boundaries to reduce step size
+        if self._lower is not None:
+            # Rectangular boundaries: reduce individual step sizes until OK
+            out = np.logical_or(p < self._lower, p >= self._upper)
```
Comment (DavAug) [inline pints/_optimisers/_irpropmin.py:246]: Is there a reason for including the upper boundary and not the lower boundary?

File: pints/_optimisers/_irpropmin.py [hunk 4]
```diff
@@ -89,6 +109,15 @@ def __init__(self, x0, sigma0=0.1, boundaries=None):
         # Current step sizes
         self._step_size = np.array(self._sigma0)
 
+        # Rectangular boundaries
+        self._lower = self._upper = None
+        if isinstance(self._boundaries, pints.RectangularBoundaries):
+            self._lower = self._boundaries.lower()
+            self._upper = self._boundaries.upper()
```
Comment (DavAug) [inline pints/_optimisers/_irpropmin.py:116]: I am wondering whether it's more clear to move this to line 244. The self._lower and self._upper are not used in any other methods are they? And to me, it's not immediately clear that line 244 is only being invoked for rectangular boundaries.

File: pints/_optimisers/_irpropmin.py [hunk 5]
```diff
@@ -178,7 +238,50 @@ def tell(self, reply):
         self._current_df = dfx
 
         # Take step in direction indicated by current gradient
-        self._proposed = self._current - self._step_size * np.sign(dfx)
+        p = self._current - self._step_size * np.sign(dfx)
+
+        # Allow boundaries to reduce step size
+        if self._lower is not None:
```
Comment (DavAug) [inline pints/_optimisers/_irpropmin.py:244]: See comment above about making more clear that this is only used for rectangular boundaries

**Review by DavAug — COMMENTED** [review]
Thanks @MichaelClerx this looks very good! I have not reviewed everything yet, but I thought I leave some comments already. I'll have another look tomorrow 😊

File: pints/_optimisers/_irpropmin.py (hunk 1 above)
Comment (MichaelClerx) [reply pints/_optimisers/_irpropmin.py:189]: 😳  I didn't add it out of laziness.

Do you think we should validate late?
I wouldn't want this situation:

```
# Set min step size to 1, max to 10
opt.set_min_step_size(1)

ERROR! MAX STEP SIZE IS 0.1
```

so I was thinking the check would happen on the first call to `ask()` ?

File: pints/_optimisers/_irpropmin.py (hunk 3 above)
Comment (MichaelClerx) [reply pints/_optimisers/_irpropmin.py:246]: Yeah, the RectangularBoundaries class uses a half-open interval: https://github.com/pints-team/pints/blob/master/pints/_boundaries.py#L91-L97

This code is doing the same check manually, so should use the same convention

File: pints/_optimisers/_irpropmin.py (hunk 4 above)
Comment (MichaelClerx) [reply pints/_optimisers/_irpropmin.py:116]: 🤔 
I want to limit the number of function calls & inspection though, so no `isinstance` or `lower()` in `tell()`.
Not a deal-breaker though. Happy to use the isinstance check inside tell if you prefer

(I'm being efficient here out of habit, not really out of design)

**Comment by MichaelClerx** [comment]
Thanks!

File: pints/_optimisers/_irpropmin.py (hunk 1 above)
Comment (DavAug) [reply pints/_optimisers/_irpropmin.py:189]: Hmm I don't have a strong opinion on this 😊 I personally think it's nice when attempting to set a hyperparameter that breaks the method immediately raises an error, or maybe at least a UserWarning? But if the error thrown in `ask()` is clear enough about which hyperparameter breaks the optimiser, I am also happy with that 😊

File: pints/_optimisers/_irpropmin.py (hunk 4 above)
Comment (DavAug) [reply pints/_optimisers/_irpropmin.py:116]: I am happy either way 😊 The way it is right now is probably easy enough to understand, but I would be surprised if the boundary method calls have a significant impact on the run time.

**Review by DavAug — APPROVED** [review]
Looks good @MichaelClerx ! I also think that the step size adjustment for non-rectangular boundaries feels a little clumsy and, as you said, there is definitely bias to move towards smaller index dimensions. But at the same time, I think this will only start to matter once we introduce more complicated boundaries than currently available in pints.

**Comment by MichaelClerx** [comment]
Thanks @DavAug ! I've added the checks and merged it in

**Pull request merged**